src/wp-includes/option.php

```diff
@@ -784,14 +784,41 @@ function update_option( $option, $value, $autoload = null ) {
 
 	/*
 	 * If the new and old values are the same, no need to update.
-	 *
+	 * Scalar values in the cache will always be strings, so we must compare string values.
+	 */
+	$values = array(
+		'old' => $old_value,
+		'new' => $value,
+	);
+
+	foreach ( $values as $_key => &$_value ) {
+		// Special handling for false-ish values.
+		if ( false === $_value ) {
+			$_value = '0';
+
+			// Empty strings in the database should be seen as equivalent to false-ish cache values.
+		} elseif ( 'old' === $_key && '' === $_value ) {
+			$_value = '0';
+
+			// Cast scalars to a string so type discrepancies don't result in cache misses.
+		} elseif ( is_scalar( $_value ) ) {
+			$_value = (string) $_value;
+		}
+	}
```

Comment:
Could this repeated logic be broken into a util function.

Reply:
Great idea, though this should definitely be marked `@access private` for now. Maybe a `is_equal_database_value( $a, $b )`, returning a boolean?

The logic here is not specific for options, it would work just as well for lots of other database values like e.g. metadata etc. So I think naming the function something generic about database values makes sense.

Reply:
We add unit tests for just this new function as well.

Reply:
Thanks for suggestions. In https://github.com/WordPress/wordpress-develop/pull/5139/commits/6497e622371cb139082bf7819b8b0014e31e907f and https://github.com/WordPress/wordpress-develop/pull/5139/commits/e5a5eac441f4fa78f6552664496316fc62f1c88a I have added a new utility function and unit tests.